### ISSUE TITLE
For a dataset of nested elements `window` creates a dataset of nested datasets of flat elements, not a dataset of datasets of nested elements.

### ISSUE THREAD

**Issue by scd75** [tensorflow/tensorflow#43703]
For a dataset of nested elements `window` creates a dataset of nested datasets of flat elements, not a dataset of datasets of nested elements.

In other words, the `x` in your `lambda` will be a dictionary mapping keys to datasets and if you wish to flatten it to a single dataset you will need to do:

```
def map_fn(x):
  result = {}
  for key in x.keys():
    result[key] = tf.data.experimental.get_single_element(x[key].batch(3))
  return result

dataset = tf.data.Dataset.from_tensor_slices({"a": [1,2,3,4]}).window(3).map(map_fn)
```

_Originally posted by @jsimsa in https://github.com/tensorflow/tensorflow/issues/23581#issuecomment-529702702_

When running the above, I am able to get a Dataset of dictionnaries of tensors of shape (window_size,).
However, it seems that batching futher such object then triggers an 'out or range' error in model training, and even when going through the elements. Such error seem to occur only when number of windows are not amultiple of the final batch size..

```

import numpy as np
import pandas as pd
import tensorflow as tf

window_size = 30

def map_fn(x):
  result = {}
  for key in x.keys():
    result[key] = tf.data.experimental.get_single_element(x[key].batch(window_size))
  return result

data = pd.DataFrame(
    np.random.random_sample((200,5)),
    columns=['feature1', 'feature2', 'feature3', 'feature4', 'feature5'],
)

dataset = tf.data.Dataset.from_tensor_slices(dict(data))
dataset = dataset.window(window_size, drop_remainder=True)
dataset = dataset.map(map_fn)

for elem in dataset:
    print(elem)

dataset = dataset.batch(32)

for elem in dataset:
    print(elem)
```

**Comment by scd75** [comment]
When running the above, I am able to get a Dataset of dictionnaries of tensors of shape (window_size,).
However, it seems that batching futher such object then triggers an 'out or range' error in model training, and even when going through the elements. Such error seem to occur only when number of windows are not amultiple of the final batch size..

```

import numpy as np
import pandas as pd
import tensorflow as tf

window_size = 30

def map_fn(x):
  result = {}
  for key in x.keys():
    result[key] = tf.data.experimental.get_single_element(x[key].batch(window_size))
  return result

data = pd.DataFrame(
    np.random.random_sample((200,5)),
    columns=['feature1', 'feature2', 'feature3', 'feature4', 'feature5'],
)

dataset = tf.data.Dataset.from_tensor_slices(dict(data))
dataset = dataset.window(window_size, drop_remainder=True)
dataset = dataset.map(map_fn)

for elem in dataset:
    print(elem)

dataset = dataset.batch(32)

for elem in dataset:
    print(elem)
```

**Comment by jsimsa** [comment]
This is working as intended so that you can for instance apply different batching logic on different elements of the nest.

**Comment by scd75** [comment]
You can try the code I wrote: it is not working, it returns "out or range" error for some window_size, batch numbers (such as those I entered).

**Comment by jsimsa** [comment]
Sorry, the title of the issue mislead me and I didn't read the details. I can confirm I can reproduce the issue and will investigate.

**Comment by jsimsa** [comment]
There was indeed a bug and I have a pending fix out which when submitted will autoclose this issue.

**Comment by scd75** [comment]
Hello, thanks for re-checking, however I installed tf-nightly today, and still have the issue when running the code above..

**Comment by ravikyram** [comment]
@scd75 

I am not seeing any issue in nightly version.Please,find the gist [here](https://colab.research.google.com/gist/ravikyram/d92fd919aaeac0f7137b15b849d46dab/untitled412.ipynb).Thanks!

**Comment by scd75** [comment]
Hi, thanks I get it.. I'm working on a windows env, and for some reason the last tf-nightly version available for windows on Pypi is 20201002. Will re-check when available.

**Comment by scd75** [comment]
so I got this method working..
My issue is that I want to create windowed datasets with named features. I actually found out that this method is super slow when running training, taking ages to fill shuffle buffer at the begining of each epoch.

I ended up trying another approach. Here is the code of the 2 approaches I coded, first one being based on the method above, 2nd one being based on an "ex-post" naming of each feature.
The filling up of buffer at the begining of each epoch is 10 times faster with the 2nd method. The main issue with the 2nd method is that it requires all features to have the same dtype, as they are processed through a signle tensor..

I think it would be very useful to have a better / cleaner way of generating window datasets with named features, while preserving speed at training time. Would appreciate if you could dig into that.. Thanks!

Following is my code for the 2 methods:

```

def make_window_dataset_tf(data, target_key, lags, horizon, shift=1, stride=1, shuffle=False, seed=None, batch_size=32):
    
    def map_fn(x):
        result = {}
        for key in x.keys():
            result[key] = tf.data.experimental.get_single_element(x[key].batch(lags+horizon))
        return result

    ds = tf.data.Dataset.from_tensor_slices(dict(data))
    ds = ds.window(size=lags+horizon, shift=shift, stride=stride,drop_remainder=True)
    ds = ds.map(map_fn)
    ds = ds.map(lambda x: (x, x[target_key][-horizon:]))
    if shuffle:
        ds = ds.shuffle(buffer_size=len(data),seed=seed)
    ds = ds.batch(batch_size)

    return ds


def make_window_dataset_tf2(data, target_key, lags, horizon, shift=1, stride=1, shuffle=False, seed=None, batch_size=32):

    def tensor_to_dict(x):
        dic = {}
        for i, key in enumerate(data):
            dic[key] = x[Ellipsis,i:i+1]
        return dic

    ds = tf.data.Dataset.from_tensor_slices(data)
    ds = ds.window(size=lags+horizon, shift=shift, stride=stride,drop_remainder=True)
    ds = ds.flat_map(lambda x: x.batch(lags+horizon))
    ds = ds.map(tensor_to_dict)
    ds = ds.map(lambda x: (x, x[target_key][-horizon:]))
    if shuffle:
        ds = ds.shuffle(buffer_size=len(data),seed=seed)
    ds = ds.batch(batch_size)

    return ds
```
